NVIM_APPNAME supports relative paths

RoryNesbitt opened the pull request:
This is a minimum viable product to allow paths in `NVIM_APPNAME`. It is enough to fix #23056 and close #24966

The current implementation means a user with multiple configs will end up with a lot of messy directories that may not be clear they relate to neovim (depending on what the appname they used once and then forgot about was).

Allowing the appname to include sub-directories would mean that they can choose to group all these together,
e.g.
```
NVIM_APPNAME="neovim-configs/first-config" nvim
NVIM_APPNAME="neovim-configs/second-config" nvim
```
It would also allow the user to clone someones dotfiles repo and then reach the config without moving it from the repo (this is a key part to how [dotfyle-cli](https://github.com/rorynesbitt/dotfyle-cli) works),
e.g.
```
git clone git@github.com:user/dotfiles ~/.config/user/dotfiles
NVIM_APPNAME="user/dotfiles/.config/nvim" nvim
```

I am looking for feedback about what else would need to be included in this for it to be merged.
I don't know c but I'm willing to do what I can if pointed in the right direction

From the relevant issue:
- ~updated stdpath logic in various places~
- [x] update to the docs
- [x] block the inclusion of `../` in the appname

RoryNesbitt committed:
feat: relative path in NVIM_APPNAME

Problem:
NVIM_APPNAME does not allow path separators in the name, so relative
paths can't be used:

    NVIM_APPNAME="neovim-configs/first-config" nvim
    NVIM_APPNAME="neovim-configs/second-config" nvim

Solution:
Let NVIM_APPNAME be a relative path. Absolute paths are not supported.

fix #23056
fix #24966

RoryNesbitt commented:
The new appname check looks for the equivalent of`/../`, `^../`, or `/..$`     (or `\\`)
This means that if someone for whatever reason has .. in the middle of a valid filename then that will still work, it will only get blocked when it is being used to mean parent directory.

This should mean that `XDG_xx` variables are still correctly made use of, but still allow for subdirectories. In which case are the other two changes still required?

justinmk commented:
Please always say clearly what problem is being solved, and give an example. 

The first obvious question is, if `NVIM_APPNAME=$HOME/t` then what are `stdpath('config')` and `stdpath('cache')` expected to return? 

> In which case are the other two changes still required?

re "updated stdpath logic in various places": for example, `stdpath('config')` and `stdpath('cache')` are currently assumed (correctly) to be separate directories. But if `NVIM_APPNAME` can be set to an absolute path then that falls apart. Neither Nvim nor plugins can reason about "the cache directory" if it's the same directory as "the config directory". 

### Other notes:

- > The new appname check looks for the equivalent of`/../`, 
    - Instead of checking for ".." literally, can the path be resolved to an unambiguous, absolute path--which is then compared with the umabiguous, absolute form of CWD or its parent? That should be more robust. Though I don't remember if we have a robust "normalize" function internally, yet.
- If this PR ends up being viable then it would need tests, e.g. the [discussion](https://github.com/neovim/neovim/discussions/24954#discussioncomment-6870419) mentions `tempname()`.

RoryNesbitt commented:
> Please always say clearly what problem is being solved, and give an example.

I've updated the original comment

> The first obvious question is, if `NVIM_APPNAME=$HOME/t` then what are `stdpath('config')` and `stdpath('cache')` expected to return?

This is an interesting question and not something I was planning to support here. The idea behind checking for `..` is that it has to remain a subdirectory of the normal `stdpath('config')` and `stdpath('cache')` etc
I suppose the extension of this logic is to also make sure the appname doesn't start with a path separator, `$HOME/t` would become `/home/user/t` and therefore the initial `/` would throw the error.

> re "updated stdpath logic in various places": for example, `stdpath('config')` and `stdpath('cache')` are currently assumed (correctly) to be separate directories. But if `NVIM_APPNAME` can be set to an absolute path then that falls apart. Neither Nvim nor plugins can reason about "the cache directory" if it's the same directory as "the config directory".

Again, this is not the intention of this pr


>  * Instead of checking for ".." literally, can the path be resolved to an unambiguous, absolute path (which is then compared with the umabiguous, absolute form of CWD or its parent)? That should be more robust. Though I don't remember if we have a robust "normalize" function internally, yet.

It was suggested to me that I make use of realpath to check the actual location of the directory. Although this would mean symlinks to strange places would not work and my personal opinion on that is if the user is symlinking they probably have a reason for it (like gnu stow). So my check deliberately stuck to the idea of "what does the string look like" not "where does the resulting directory rest".
I am of course open to differing opinions from the more wise

> * If this PR ends up being viable then it would need tests, e.g. the [discussion](https://github.com/neovim/neovim/discussions/24954#discussioncomment-6870419) mentions `tempname()`.

I shall make sure to add tests. `tempname()` should be the only thing really effected,,I think

RoryNesbitt commented:
> The first obvious question is, if NVIM_APPNAME=$HOME/t then what are stdpath('config') and stdpath('cache') expected to return?

After some testing I don't think this is actually a problem (although I would still block it because it could cause confusion).

When I set the appname to this I don't get anything put in `/home/rory/t` I get `/home/rory/.local/state/home/rory/t` etc.
I am inferring from this that an absolute path is never actually possible and since linux treats `//` like `/` it will "just work". But I would not expect it to just work on windows, in fact I don't even know if the check I added would work on windows

RoryNesbitt commented:
I'm not entirely sure if that's the best way to do the tests, I based them on the one that was already there.
Also the old test suggests that `\\` should be read as a `PATHSEP` but I'm assuming that would only be windows? in which case maybe I should seperate the tests with `if is_os('win') then` and test for both `/` and `\\` options

RoryNesbitt commented:
Hey @justinmk , I think I've addressed all your issues, would appreciate another look when you get the time.

Things I'm anticipating, but I'm not sure about:
- Move the updated test that makes sure incorrect appnames are ignored to a separate test, rather than the one it was in before when it was a simpler check? If so where would you want it?
- rebase the commits to be shorter, or would you squash them all together? Maybe something like 
- - fix(tempname): substitute PATHSEP for '_' in directory name
- - feat(NVIM_APPNAME): allow relative paths
- - test(options/defaults_spec): add tests for NVIM_APPNAME as a path

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,13 +69,19 @@ const char *get_appname(void)
   const char *appname = get_appname();
   const size_t appname_len = strlen(appname);
+  if (appname[0] == PATHSEP) {
+    return false;
```
`path_is_absolute`

RoryNesbitt replied:
I'm not sure what you're meaning here, do you want me to set that variable instead of the early return?

justinmk replied:
i'm saying `path_is_absolute` should be used instead of checking `appname[0] == PATHSEP`

RoryNesbitt replied:
I did not know that was an option, that would have saved such a headache yesterday.
Thanks for the info :smile: updated it

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,13 +69,19 @@ const char *get_appname(void)
+  if (appname[0] == PATHSEP) {
+    return false;
+  }
   for (size_t i = 0; i < appname_len; i++) {
```
is the for-loop still needed?

RoryNesbitt replied:
The for loop is specifically for `..` as although absolute paths can't be used you can do the equivalent by setting appname to `../../../../../effectively/absolute/path`

RoryNesbitt replied:
I double checked just in case, the new test will fail if the for loop is taken out

justinmk commented on test/functional/options/defaults_spec.lua:
```diff
@@ -616,10 +616,25 @@ describe('stdpath()', function()
-      local child = vim.fn.jobstart({ vim.v.progpath }, { env = { NVIM_APPNAME = 'a/b\\c' } })
-      return vim.fn.jobwait({ child }, 3000)[1]
-    ]]))
+    local function runAppnameTest(testAppname, expectedExitCode)
```
```suggestion
    local function test_appname(testAppname, expectedExitCode)
```

justinmk commented:
> as per the ci checks, would you want something about this in news?

yes. also `:help NVIM_APPNAME` will need to say "must be a name or relative path"

> local waitTime = (expectedExitCode == -1) and 100 or 3000 This works, but do we actually care about wasting time in tests?

test time matters very much, we'll need to find a better approach

RoryNesbitt commented:
> yes. also `:help NVIM_APPNAME` will need to say "must be a name or relative path"

Added both these things

> test time matters very much, we'll need to find a better approach

The idea behind that line of code was that if we expect nvim to error with that appname set then it would happen very quickly anyway, and 3000 was the timeout used on the previous check incase it didn't. However, if we expect nvim to accept the appname then we will always be waiting for the timeout to occur, so I made it much shorter. 
I fear to say "it works on my machine" but an alternative option is to keep the timeout always at 100. My tests still get the correct results with that set

RoryNesbitt commented:
I done the rebase to appease the commit linter. Hope that's chill :smile:

justinmk commented on runtime/doc/news.txt:
```diff
@@ -169,6 +169,8 @@ The following new APIs and features were added.
 • |vim.lsp.util.locations_to_items()| sets the `user_data` of each item to the
   original LSP `Location` or `LocationLink`.
 
+• |$NVIM_APPNAME| environment variable now supports being set to a relative path
```
```suggestion
• |$NVIM_APPNAME| can be set to a relative path instead of only a name.
```

justinmk commented on src/nvim/fileio.c:
```diff
@@ -3309,7 +3309,13 @@ static void vim_mktempdir(void)
+    char appname[40] = { 0 };
+    xstrlcpy(appname, get_appname(), sizeof(appname));
+    // Substitute slashes in appname with underscores
+    memchrsub(appname, '/', '_', sizeof(appname));
+    memchrsub(appname, '\\', '_', sizeof(appname));
```
Should we use `%` since that is what is used for other things like undofiles ? Could also reduce the chances of accidental conflicts.

```suggestion
    memchrsub(appname, '/', '%', sizeof(appname));
    memchrsub(appname, '\\', '%', sizeof(appname));
```

RoryNesbitt replied:
That makes sense, it's currently `_` because that's the same change that's made to [name, just above](https://github.com/neovim/neovim/blob/master/src/nvim/fileio.c#L3294-L3297).
If we were to change this here I would suggest changing both

justinmk commented on src/nvim/fileio.c:
```diff
@@ -3309,7 +3309,13 @@ static void vim_mktempdir(void)
 
-    const char *appname = get_appname();
+    // Make a copy of the appname to avoid modifying it
+    char appname[40] = { 0 };
```
where is 40 coming from?

RoryNesbitt replied:
Similar to the above comment, I heavily referenced the way name worked when writing this 
[Seen here](https://github.com/neovim/neovim/blob/master/src/nvim/fileio.c#L3292)

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,13 +69,19 @@ const char *get_appname(void)
 }
 
-/// Ensure that APPNAME is valid. In particular, it cannot contain directory separators.
+// Ensure that APPNAME is valid. In particular, it cannot contain .. surrounded
+// by path separators, the start of the string, or the end of the string.
```
No reason to allow ".." at all. That simplifies both the documentation and the logic.

```suggestion
// Ensure that APPNAME is valid. Must be a name or relative path.
```

RoryNesbitt replied:
What if a user has a directory name like `nvim..configs`? It's weird, I don't like it, but it's a valid dir name

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,13 +69,19 @@ const char *get_appname(void)
+    if (appname[i] == '.' && appname[i + 1] == '.'
+        && (i == 0 || appname[i - 1] == PATHSEP)
+        && (appname[i + 2] == '\0' || appname[i + 2] == PATHSEP)) {
       return false;
```
Can we get rid of the for-loop, and just return false if `..` is found anywhere in the string?

RoryNesbitt replied:
This makes sense but would fail for instances of `..dir` or `dir..`. Although these must be very niche cases so I'd be comfortable overlooking them

RoryNesbitt replied:
To cover those I could check for 
```
/../
^../
/..$
^..$
```
This is pretty much what the for loop checks for but would be clearer

justinmk replied:
NVIM_APPNAME is a completely controllable value, there's no reason we need to support weird cases. Better for the logic to be simpler where possible.

RoryNesbitt replied:
Took me way too long to realise I could use PATHSEPSTR to make this just work.

I also added a check for APPNAME == "." I feel that would break things

justinmk commented on test/functional/options/defaults_spec.lua:
```diff
@@ -914,3 +929,14 @@ describe('stdpath()', function()
 end)
+
+describe('tempname()', function()
+  describe('returns a string', function()
```
don't need a new describe() for this, it should be sibling to the other test above

RoryNesbitt replied:
I'm not sure how to word the check but I've moved it into place

RoryNesbitt replied:
Took me way too long to realise I could use `PATHSEPSTR` to make this just work.

I also added a check for `APPNAME == "."` I feel that would break things

justinmk commented on src/nvim/fileio.c:
```diff
@@ -3309,7 +3309,13 @@ static void vim_mktempdir(void)
+    // Make a copy of the appname to avoid modifying it
+    char appname[40] = { 0 };
+    xstrlcpy(appname, get_appname(), sizeof(appname));
+    // Substitute slashes in appname with underscores
```
```suggestion
    // Replace slashes, in case appname is a relative path.
```

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,15 +69,16 @@ const char *get_appname(void)
 }
 
-/// Ensure that APPNAME is valid. In particular, it cannot contain directory separators.
+// Ensure that APPNAME is valid. Must be a name or relative path.
```
```suggestion
/// Ensure that APPNAME is valid. Must be a name or relative path.
```

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,15 +69,16 @@ const char *get_appname(void)
-      return false;
-    }
+  if (path_is_absolute(appname)
+      || strcmp(appname, ".") == 0
+      || strcmp(appname, "..") == 0
```
```suggestion
      || strequal(appname, ".")
      || strequal(appname, "..")
```

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,15 +69,16 @@ const char *get_appname(void)
+  if (path_is_absolute(appname)
+      || strcmp(appname, ".") == 0
+      || strcmp(appname, "..") == 0
+      || strstr(appname, PATHSEPSTR"..") != NULL
+      || strstr(appname, ".."PATHSEPSTR) != NULL) {
```
unfortunately this fails on windows bc of the way PATHSEPSTR is defined: https://github.com/neovim/neovim/blob/5ea5812714bcbeb6a5438c80d410984ad4fbb922/src/nvim/ascii.h#L81

let's just do this for now (both "/" and "\\" need to be checked on Windows, unless we "normalize" appname before this code is reached):

```suggestion
#ifdef BACKSLASH_IN_FILENAME
      || strstr(appname, "\\..") != NULL
      || strstr(appname, "..\\") != NULL)
#endif
      || strstr(appname, "/..") != NULL
      || strstr(appname, "/..") != NULL) {
```

justinmk commented on test/functional/options/defaults_spec.lua:
```diff
@@ -616,10 +616,33 @@ describe('stdpath()', function()
+      clear({env={
+        NVIM_APPNAME=alter_slashes('a/b'),
+      }})
+      funcs.tempname() -- should not crash
```
use `assert_alive()` after this, or ideally check for the actual value (the above test already has an approach to do that)

RoryNesbitt replied:
The thing that's throwing me with this is that my test gives it `/home/rory/Documents/projects/neovim/build/Xtest_tmpdir/a%b.rory/J4qGVP/0`
Is there a variable for `Xtest_tmpdir` that I just can't find?

RoryNesbitt replied:
Okay so I fixed that by setting `TMPDIR`, but setting `USER` did nothing, I for sure thought `USER` was the obvious one
I'll try fix it in a bit

RoryNesbitt commented on test/functional/core/fileio_spec.lua:
This test works better than the previous, but I have to be honest I don't know if it's the best option, It's based on the surrounding tests

RoryNesbitt replied:
I somehow put it in the wrong place when I rebased but the comment still remains

RoryNesbitt commented:
Hey @justinmk , hope you had a good weekend 😊 

I think, if you're happy with the new test, that this is ready to merge, would appreciate your opinion on that.

I've also made the substitution change further up on a different branch (`_` to `%` for the username) and run the tests, all of which came back successful.
I don't think there's a need in PRing that one, though, since it's literally changing 2 characters only, unless you want another test added to cover that too

justinmk commented:
Pushed an update.

RoryNesbitt commented:
Had a look through your changes, it all makes sense and the `matches` check is more like what I wanted but didn't know what to use :)
The review about hard coding 3000 is not something I have a strong opinion on one way or the other, just thought it should be noted.

With that all being said I am happy with the PR when you are

justinmk commented:
CI failures are unrelated. asan failure is https://github.com/neovim/neovim/issues/23762

justinmk commented on src/nvim/os/stdpaths.c:
```diff
@@ -69,15 +69,23 @@ const char *get_appname(void)
+  if (path_is_absolute(appname)
+      // TODO(justinmk): on Windows, path_is_absolute says "/" is NOT absolute. Should it?
+      || strequal(appname, "/")
+      || strequal(appname, "\\")
```
`path_is_absolute` should possibly be changed to consider "/" absolute on Windows, but not sure what the blast radius is so not fixing it in this PR.

dundargoc commented:
let's goooo!

RoryNesbitt commented:
Thanks for the help justin. Really happy to see this feature officially supported now :smile:
